library/src/hcc_detail/hipsolver.cpp

```diff
@@ -3353,53 +3464,38 @@ try
         }
 
         CHECK_ROCBLAS_ERROR(rocblas_set_workspace((rocblas_handle)handle, work, lwork));
-
-        return rocblas2hip_status(rocsolver_sgesvd((rocblas_handle)handle,
-                                                   char2rocblas_svect(jobu),
-                                                   char2rocblas_svect(jobv),
-                                                   m,
-                                                   n,
-                                                   A,
-                                                   lda,
-                                                   S,
-                                                   U,
-                                                   ldu,
-                                                   V,
-                                                   ldv,
-                                                   rwork,
-                                                   rocblas_outofplace,
-                                                   devInfo));
     }
     else
     {
         CHECK_HIPSOLVER_ERROR(
             hipsolverSgesvd_bufferSize((rocblas_handle)handle, jobu, jobv, m, n, &lwork));
         CHECK_ROCBLAS_ERROR(hipsolverManageWorkspace((rocblas_handle)handle, lwork));
 
-        rocblas_device_malloc mem((rocblas_handle)handle, sizeof(float) * min(m, n));
         if(!rwork && min(m, n) > 1)
         {
-            if(!mem)
+            mem = std::make_unique<rocblas_device_malloc>((rocblas_handle)handle,
+                                                          sizeof(float) * min(m, n));
+            if(!mem->operator bool())
                 return HIPSOLVER_STATUS_ALLOC_FAILED;
-            rwork = (float*)mem[0];
+            rwork = (float*)(*mem)[0];
         }
```

Comment:
This awkwardness exists because `rocblas_device_malloc` does not support the move semantics you want, so you're using `std::unique_ptr` which does. However, the purpose of `rocblas_device_malloc` is basically the same as `std::unique_ptr`, so we might as well just add the desired move feature to `rocblas_device_malloc`.

Reply:
I think the idea of the original design was to always initialize the resources when created (RAII), so to avoid this kind of lazy initialization with `unique_ptr` and all that. That's why I say it would be cleaner to deal with null E in rocsolver, but any option is fine by me (I don't know much about these nuances of the language anyways)... if you want to modify the `rocblas_device_malloc` class I think it would be ok; rocblas does not really use it. 

I have a question, though: in hipsolver code, as it is right now, the `rocblas_device_malloc` object `mem` only lives within the `else` statement, so it can be initialized when created (as expected). What is the problem with this approach? why do we want to change it? Is it just because you have to write twice the call to the rocsolver function (once inside the `if` and once inside the `else`)?

Reply:
> Is it just because you have to write twice the call to the rocsolver function (once inside the `if` and once inside the `else`)?

Essentially, yes. It hasn't really been a big issue before, but with gesvdj having an additional call to rocblas_geam, these copied function calls are starting to occupy many lines of code and look rather untidy.

I like @cgmb's idea of adding the relevant features to `rocblas_device_malloc` so we can drop the use of `unique_ptr`. Combined with these changes, the workspace code should look a lot more tidy.